Feature: show possible gaits

Olavhaasie opened the pull request:
Closes PM-392

## Description
This PR introduces showing which gaits are possible to be executed, a gait process status and showing the current gait. The gaits that can be currently executed are indicated with a green border and others with a red border. The style of the buttons gets updated when a gait has been finished or can be updated manually with the refresh button. Some non-gait buttons, like stop and pause, are always enabled.

_Small note:_ I still sometimes get errors that a stylesheet could not be parsed and very rarely the the input device just shuts down, with no kind of stack trace. So if you get this, describe what you did, so I can try to reproduce it. To me it seemed random.
 
## Changes
* Add status bar that shows current gait status, current gait and a _Refresh_ button
* Add stylesheet to _ui_ file for enabled and disabled buttons
* Fix `check_string` method to split on every third word to avoid to long lines in a button

Commit:
Add showing possible gaits

Commit:
Fix image button style

Commit:
Add showing label with current gait

Commit:
Fix splitting strings on multiple lines

Commit:
Allow some buttons to be always enabled

Commit:
Add refresh button

Commit:
Add updating content stylesheet

Joris-JW commented:
The concept looks nice with the coloring. However, I do get a couple of error. The first error occurred after I used the home stand, the IPD just closed randomly and the terminal stated the connection has been lost (I've got screenshots if u are interested). The second error is as u describe in the PR description. Sometimes after a movement the stylesheet could not be parsed error occurs and the GUI loses some icons and colors. It happens to me pretty often but I have not found any consistency in the events leading up to the error.

Commit:
Merge branch 'develop' into feature/PM-392-show-possible-gaits

Commit:
Merge branch 'develop' into feature/PM-392-show-possible-gaits

Commit:
Add image button

Commit:
Add graying out image buttons when disabled

Commit:
Fix styling for text buttons

Commit:
Update buttons when gait has been accepted

Commit:
Ignore paintEvent name

Olavhaasie commented:
**Second attempt**

I have changed the way the input device draws its buttons with images. I have created a new class that draws a `QPixmap`, which is a lot more efficient than setting the background image using a stylesheet. I have also made the buttons gray out when they are disabled and changed the style of the text buttons to look similar to the images. So, no more weird glitches :tada:. I did remove the colored borders, since that was no longer doable using the new button implementation, maybe we can do something similar later.

![image](https://user-images.githubusercontent.com/23523462/79758538-cfb11280-831d-11ea-88ca-0bf1825f957b.png)

Joris-JW commented:
Works much better! looks nice too. Two question though, Sometimes the refresh does not work? I'm not sure why, is this explainable? or still a bit random. Second is that in some cases the screen moves all the way down after a button is pressed, why is that? (and my thought were to maybe reverse count the lay out so the stop and emergency button are always within direct reach.)

Roelemans commented on march_rqt_input_device/src/march_rqt_input_device/input_device_view.py:
```diff
@@ -20,18 +22,32 @@ def __init__(self, ui_file, controller):
             self.create_button('rocker_switch_down', image_path='/rocker_switch_down.png',
-                               callback=lambda: self._controller.publish_decrement_step_size())
+                               callback=lambda: self._controller.publish_decrement_step_size(),
+                               always_enabled=True)
```
Why are these always enabled? I think they should only be enabled during walk at the moment.

Olavhaasie replied:
I have made these always enabled, because currently it is not possible to request from the statemachine if it is possible to increment or decrement step size. I could check for walking, since it only works during walk, but it might be nicer to create an interface to the state machine for this. Then you can also add cases where you can only increment or decrement.

Roelemans reviewed (changes requested):
Very nice :) I really like how this works :D Two small questions:
1. At startup all buttons are disabled (except those that are always_enabled), even home sit/stand. Why is that?
2. See comment
Good job though :)

Olavhaasie commented:
> Sometimes the refresh does not work? I'm not sure why, is this explainable? or still a bit random. 

Is this when pressing the _Refresh_ button or the automatic refresh? When you launch all the nodes, if the input device is launched before the state machine then the service is not yet available so all buttons will be disabled.

> Second is that in some cases the screen moves all the way down after a button is pressed, why is that?

I went to check whey you said this and it seems to happen when you are scrolled some way down and press a button. I will look into this because it is really annoying.

> At startup all buttons are disabled (except those that are always_enabled), even home sit/stand. Why is that?

Same answer I gave to the first question. When the input device launches before the state machine the service is not yet available. If you try to launch the simulation before the ipd you will see that it correctly enables the home buttons.

Joris-JW commented:
> Is this when pressing the Refresh button or the automatic refresh? When you launch all the nodes, if the input device is launched before the state machine then the service is not yet available so all buttons will be disabled.

I should have clarified my question; sometimes the _automatic refresh_ does not work after a gait is finished or selected. I'm not sure in which circumstances this happens but will test some more to see if I can pinpoint specific situation.

Commit:
Fix automatic scrolling

Roelemans reviewed (approved):
Nice :)
